### ISSUE TITLE
Building the Laravel service container from scratch

### ISSUE THREAD

**Issue by jessarcher** [BrisPHP/meetups#51]
I've been fascinated by Laravel's service container and how useful it is for quite a while now. I enjoy teaching others about it, including most recently in an episode of my podcast with JMac (of Laravel Shift fame). In that episode we talked about how a primitive service container implementation could work, and I've often thought it would be interesting to actually build a basic implementation of a service container, roughly following Laravel's API, to demonstrate and demystify it for visual/practical learners.

In 2021 this came to fruition when built a service container from scratch live at the Brisbane Laravel Meetup in January and it seems to have been very popular on the live stream and subsequent recording. I've been asked whether I would give the talk for BrisPHP and I would be happy to, so long as the audience is likely not to have already seen the talk.

Basically I start with a pre-written test suite in a vanilla PHP project, and by using TDD I build out all of the features one at a time exploring what the container can do and how it works behind the scenes. I cover basic binding and IoC, mocking, the singleton pattern, recursive dependency injection, and facades.

**Comment by ndench** [comment]
This sounds like a great talk @jessarcher! I think that most of the audience will be different so that should be fine. We'll lock it in for the [next meetup](https://www.meetup.com/BrisPHP/events/276014207/) on February 23.

**Comment by ndench** [comment]
I'll be doing this talk at the next meetup on [Februray 23](https://www.meetup.com/BrisPHP/events/276014207/).

**Comment by ndench** [comment]
Thanks for the talk at the last meetup @jessarcher! If you have any resources (slides, a github repo, etc) could you post a link here please?

**Comment by jessarcher** [comment]
Sure! The GitHub repo is https://github.com/jessarcher/service-container-from-scratch/ and there were no slides as it was all in Vim, but I think it was recorded if we have a link to that?

**Comment by ndench** [comment]
Ah yes, right you are. [Here's the YouTube video](https://www.youtube.com/watch?v=cO_39A0Y37I).
